python/ray/tune/logger/json.py

```diff
@@ -107,17 +120,38 @@ def log_trial_result(self, iteration: int, trial: "Trial", result: Dict):
         self._trial_files[trial].write("\n")
         self._trial_files[trial].flush()
 
+        self._process_launcher.launch_if_needed(
+            _upload_to_fs_path,
+            dict(
+                local_path=self._tempdir.as_posix(),
+                fs=trial.storage.storage_filesystem,
+                fs_path=trial.storage.experiment_fs_path,
+            ),
+        )
+
```

Comment:
change 2: logfiles (`result.json`) are uploaded by the callbacks themselves + cleaned up on experiment end.

The current implementation just uses a temp directory that is a local version of the experiment dir (with a bunch of trial subdirs), just for this json logger callback. Every `sync_period`, we upload the contents of this entire directory to storage.

It's unclear whether we even need frequency-based syncing here -- can we just upload on experiment end? What about ungraceful run termination?

An alternative to this method of keeping a separate temp dir for every callback is using a common ray logs folder that *doesn't* get cleaned up. See change 3.